scrapy/robotstxt.py

```diff
@@ -110,3 +110,32 @@ def allowed(self, url, user_agent):
         user_agent = to_unicode(user_agent)
         url = to_unicode(url)
         return self.rp.is_allowed(user_agent, url)
+
+
+class ProtegoRobotParser(RobotParser):
+    def __init__(self, robotstxt_body, spider):
+        from protego import Protego
+        self.spider = spider
+        try:
+            robotstxt_body = robotstxt_body.decode('utf-8')
+        except UnicodeDecodeError:
+            # If we found garbage or robots.txt in an encoding other than UTF-8, disregard it.
+            # Switch to 'allow all' state.
+            logger.warning("Failure while parsing robots.txt using %(parser)s."
+                           " File either contains garbage or is in an encoding other than UTF-8, treating it as an empty file.",
+                           {'parser': "ProtegoRobotParser"},
+                           exc_info=sys.exc_info(),
+                           extra={'spider': self.spider})
```

Comment:
given that this code is repeated several times, I think it makes sense to create a function for it